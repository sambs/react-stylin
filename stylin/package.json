{
  "name": "@sambs/react-stylin",
  "version": "0.1.0",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/esm/index.d.ts",
  "files": ["dist"],
  "peerDependencies": {
    "react": "^16.13.1"
  }
}
