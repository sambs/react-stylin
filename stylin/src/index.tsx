import * as React from 'react'

export type StyleResolver<T, P extends StyleProps> = (params: {
  theme: T
  props: P
  defaults: React.CSSProperties
}) => React.CSSProperties

export function useStylin<T, P extends StyleProps>(
  themeContext: React.Context<T>,
  props: P,
  ...resolvers: Array<StyleResolver<T, P> | undefined>
) {
  const theme = React.useContext(themeContext)

  let style: React.CSSProperties = {}

  for (const resolver of resolvers) {
    if (resolver) {
      style = resolver({ theme, props, defaults: style })
    }
  }
  return style
}

type ElementName = keyof JSX.IntrinsicElements

type StyleProps = { [prop: string]: any }

type StylinElementProps<
  E extends ElementName,
  T,
  P extends StyleProps
> = JSX.IntrinsicElements[E] & {
  styles?: StyleResolver<T, P>
} & Partial<P>

type CreateStylinElementParams<
  E extends ElementName,
  T,
  P extends StyleProps
> = {
  defaultProps?: JSX.IntrinsicElements[E]
  defaultStyles?: StyleResolver<T, P>
  defaultStyleProps?: P
  displayName?: string
  element: E
  themeContext: React.Context<T>
}

export const createStylinElement = <
  E extends ElementName,
  T,
  P extends StyleProps
>({
  defaultProps,
  defaultStyles,
  defaultStyleProps,
  displayName = 'Element',
  element,
  themeContext,
}: CreateStylinElementParams<E, T, P>) => {
  const component: React.FC<StylinElementProps<E, T, P>> = (props) => {
    const styleProps: Partial<P> = {}
    const elemProps: JSX.IntrinsicElements[E] = {}

    for (const prop in defaultStyleProps) {
      if (prop in props) {
        styleProps[prop] = props[prop]
      } else {
        styleProps[prop] = defaultStyleProps[prop]
      }
    }

    for (const prop in props) {
      if (defaultStyleProps && defaultStyleProps.hasOwnProperty(prop)) continue
      if (prop == 'styles') continue
      // @ts-ignore
      elemProps[prop] = props[prop]
    }

    const style = useStylin(
      themeContext,
      styleProps as Required<P>,
      defaultStyles,
      props.styles
    )

    return React.createElement(element, {
      ...defaultProps,
      ...elemProps,
      style,
    })
  }

  component.displayName = displayName

  return component
}
