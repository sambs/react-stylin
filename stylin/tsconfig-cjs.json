{
  "extends": "./tsconfig-esm.json",
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "outDir": "./dist/cjs"
  },
  "include": ["./src/**/*"]
}
