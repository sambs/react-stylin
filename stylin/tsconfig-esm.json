{
  "extends": "../tsconfig-base.json",
  "compilerOptions": {
    "module": "es2020",
    "declaration": true,
    "outDir": "./dist/esm",
    "rootDir": "./src"
  },
  "include": ["./src/**/*"]
}
