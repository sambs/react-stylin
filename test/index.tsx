import * as React from 'react'
import { createStylinElement } from 'stylin'
import { create as createRenderer } from 'react-test-renderer'

export const theme = {
  colors: {
    primary: 'red',
    secondary: 'green',
  },
  text: {
    lg: {
      fontSize: '24px',
    },
    md: {
      fontSize: '16px',
    },
    sm: {
      fontSize: '12px',
    },
  },
}

type Theme = typeof theme

const ThemeContext = React.createContext(theme)

const createRendererWithContext = (component: JSX.Element) =>
  createRenderer(
    <ThemeContext.Provider value={theme}>{component}</ThemeContext.Provider>
  )

const Alert = createStylinElement({
  element: 'div',
  displayName: 'Alert',
  defaultProps: { role: 'alert' },
  defaultStyles: ({ theme }) => ({
    background: theme.colors.primary,
  }),
  themeContext: ThemeContext,
})

test('Custom Element with default props', () => {
  const component = createRendererWithContext(<Alert>Uh Oh!</Alert>)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

type TextStyleProps = {
  size: 'sm' | 'md' | 'lg'
}

const Text = createStylinElement<'div', Theme, TextStyleProps>({
  element: 'div',
  displayName: 'Text',
  defaultStyleProps: {
    size: 'md',
  },
  defaultStyles: ({ theme, props }) => theme.text[props.size],
  themeContext: ThemeContext,
})

test('Custom Element with style props', () => {
  let component = createRendererWithContext(<Text>Hey there!</Text>)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  component = createRendererWithContext(<Text size="sm">Hey there!</Text>)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  component = createRendererWithContext(
    <Text styles={() => ({ fontSize: '15px' })}>Hey there!</Text>
  )
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
