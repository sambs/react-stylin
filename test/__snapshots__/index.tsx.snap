// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Element with default props 1`] = `
<div
  role="alert"
  style={
    Object {
      "background": "red",
    }
  }
>
  Uh Oh!
</div>
`;

exports[`Custom Element with style props 1`] = `
<div
  style={
    Object {
      "fontSize": "16px",
    }
  }
>
  Hey there!
</div>
`;

exports[`Custom Element with style props 2`] = `
<div
  style={
    Object {
      "fontSize": "12px",
    }
  }
>
  Hey there!
</div>
`;

exports[`Custom Element with style props 3`] = `
<div
  style={
    Object {
      "fontSize": "15px",
    }
  }
>
  Hey there!
</div>
`;
